### ISSUE TITLE
Allow restoring subgraph from checkpoint for weight sharing between different models

### ISSUE THREAD

**Issue by tremblerz** [tensorflow/tensorflow#20608]
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS Sierra
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.7.0-3-g024aecf414 1.7.0
- **Python version**: 3
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: Read the summary below

### Describe the problem
There are many scenarios where we want to do parameter sharing between two different models (let's call `A` and `B`). If `A` is trained and checkpointed, we can either load all of the variables of `A` or selectively load variables from A's checkpoint by passing a list `L` containing all shared params. But the second way is not helpful if we don't know `L` beforehand. Can we have a way by which we can share variables just based on the common subset between two graphs.
PS - I am assuming that parameters of `A` and `B` follow same naming convention in scopes.

### Source code / logs
Here is what I was trying to do using metagraph but this also doesn't seem to work. Please note that I deliberately do `tf.reset_default_graph()` as objective is to share parameter with a checkpointed model. Sorry lots of redundant lines in the code.
```
import tensorflow as tf
mnist = tf.contrib.learn.datasets.load_dataset("mnist")

def model_1(features, labels):
    with tf.variable_scope("input", reuse=tf.AUTO_REUSE):
        out = tf.reshape(features, [-1, 28, 28, 1])
    with tf.variable_scope("conv_1_5x5", reuse=tf.AUTO_REUSE):
        out = tf.layers.conv2d(
           inputs=out,
           filters=32,
           kernel_size=[5, 5],
           padding="same",
           activation=tf.nn.relu)
    with tf.variable_scope("conv_2_3x3", reuse=tf.AUTO_REUSE):
        out = tf.layers.conv2d(
           inputs=out,
           filters=32,
           kernel_size=[3, 3],
           padding="same",
           activation=tf.nn.relu)
    with tf.variable_scope("pool_3_2x2", reuse=tf.AUTO_REUSE):
        out = tf.layers.max_pooling2d(inputs=out, pool_size=[2, 2], strides=2)
    with tf.variable_scope("conv_4_3x3", reuse=tf.AUTO_REUSE):
        out = tf.layers.conv2d(
           inputs=out,
           filters=64,
           kernel_size=[3, 3],
           padding="same",
           activation=tf.nn.relu)
    with tf.variable_scope("dense", reuse=tf.AUTO_REUSE):
        flattened = tf.reshape(out, [-1, 14 * 14 * 64])
        logits = tf.layers.dense(flattened, units=10, activation=tf.nn.relu)
    softmax = tf.nn.softmax(logits, name="softmax")

    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
    optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001)
    train_op = optimizer.minimize(loss=loss, global_step=tf.train.get_or_create_global_step())
    return loss, train_op

def model_2(features, labels, old_graph):
    with old_graph.as_default():
        with tf.variable_scope("input", reuse=tf.AUTO_REUSE):
            out = tf.reshape(features, [-1, 28, 28, 1])
        with tf.variable_scope("conv_1_5x5", reuse=tf.AUTO_REUSE):
            out = tf.layers.conv2d(
               inputs=out,
               filters=32,
               kernel_size=[5, 5],
               padding="same",
               activation=tf.nn.relu)
        with tf.variable_scope("conv_2_3x3", reuse=tf.AUTO_REUSE):
            out = tf.layers.conv2d(
               inputs=out,
               filters=32,
               kernel_size=[3, 3],
               padding="same",
               activation=tf.nn.relu)
        with tf.variable_scope("pool_3_2x2", reuse=tf.AUTO_REUSE):
            out = tf.layers.max_pooling2d(inputs=out, pool_size=[2, 2], strides=2)
        # This is the place where model 1 and 2 are actually different
        with tf.variable_scope("conv_4_5x5", reuse=tf.AUTO_REUSE):
            out = tf.layers.conv2d(
               inputs=out,
               filters=64,
               kernel_size=[5, 5],
               padding="same",
               activation=tf.nn.relu)
        with tf.variable_scope("dense", reuse=tf.AUTO_REUSE):
            flattened = tf.reshape(out, [-1, 14 * 14 * 64])
            logits = tf.layers.dense(flattened, units=10, activation=tf.nn.relu)
        softmax = tf.nn.softmax(logits, name="softmax")

    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
    optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001)
    train_op = optimizer.minimize(loss=loss, global_step=tf.train.get_or_create_global_step())
    return loss, train_op

features = tf.placeholder(shape=[None, 784], dtype=tf.float32, name="features")
labels = tf.placeholder(shape=[None], dtype=tf.int32, name="labels")

steps = 0
loss_op, train_op = model_1(features, labels)

with tf.train.MonitoredTrainingSession(checkpoint_dir="./model_1") as sess:
    while steps <= 20:
        features_, labels_ = mnist.train.next_batch(100)
        loss, _ = sess.run([loss_op, train_op], feed_dict={
            "features:0": features_, "labels:0": labels_})
        if steps % 5 == 0:
            print("loss at step {} = {}".format(steps, loss))
        steps += 1

tf.reset_default_graph()
tf.train.import_meta_graph("./model_1/model.ckpt-"+str(steps)+".meta")
old_graph = tf.get_default_graph()

steps = 0
loss_op, train_op = model_2(features, labels, old_graph)

with tf.train.MonitoredTrainingSession(checkpoint_dir="./model_2") as sess:
    while steps <= 20:
        features_, labels_ = mnist.train.next_batch(100)
        loss, _ = sess.run([loss_op, train_op], feed_dict={
            "features:0": features_, "labels:0": labels_})
        if steps % 5 == 0:
            print("loss at step {} = {}".format(steps, loss))
        steps += 1

**Comment by asimshankar** [comment]
If you can organize the "subgraphs" into `tf.keras.Model` objects, then object-based checkpointing ([`tf.train.Checkpoint` in TF 1.9](https://www.tensorflow.org/versions/r1.9/api_docs/python/tf/train/Checkpoint)) is hopefully the most intuitive.

However, this is something that's in the early stages of being exposed and we haven't finished integration with `MonitoredSession` etc. just yet. @allenlavoie will have updates on that end, so assigning to him.

As an example of how things might look:

```python
import tensorflow as tf

L = tf.keras.layers # Shorthand

def make_shared_model():
    return tf.keras.Sequential([
        L.Conv2D(filters=32, kernel_size=(5, 5), padding='same', activation=tf.nn.relu, input_shape=[28, 28, 1]),
        L.Conv2D(filters=32, kernel_size=(3,3), padding='same', activation=tf.nn.relu),
        L.MaxPool2D(pool_size=(2,2), strides=2)])

def make_model1():
    return tf.keras.Sequential([
        make_shared_model(),
        L.Conv2D(filters=64, kernel_size=(3, 3), padding='same', activation=tf.nn.relu),
        L.Flatten(),
        L.Dense(10)])

model1 = make_model1()
# Create a checkpoint with just the shared portion, which is model1.layers[0]
ckpt1 = tf.train.Checkpoint(m = model1.layers[0])

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for step in range(20):
        # Get your data and train model1
        # Save a checkpoint of the shared components
        pass
    ckpt1.save("/tmp/checkpoint")


# Could have done all this in a single graph, but demonstrating two graphs
# just to be parallel to the code you had above
tf.reset_default_graph()

def make_model2():
    return tf.keras.Sequential([
        make_shared_model(),
        L.Conv2D(filters=64, kernel_size=(5, 5), padding='same', activation=tf.nn.relu),
        L.Flatten(),
        L.Dense(10)])

model2 = make_model2()
# Since model1.layers[0] and model2.layers[0] are isomorphic objects, they can
# save and load checkpoints compatible with each other
ckpt2 = tf.train.Checkpoint(m = model2.layers[0])
status = ckpt2.restore("/tmp/checkpoint")

with tf.Session() as sess:
  status.initialize_or_restore()
  # At this point, the weights for the common portion have been restored
```

Hope this is helpful while we make object based checkpointing more pervasive.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 119 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 134 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@tremblerz,
This can be easily achieved in **`Tensorflow 2.x`** using [tf.keras.model.save_weights](https://www.tensorflow.org/api_docs/python/tf/keras/Model#save_weights) and then reuse those **`Weights`** by **`Loading the Weights`** using [tf.keras.model.load_weights](https://www.tensorflow.org/api_docs/python/tf/keras/Model#load_weights). Thanks!
